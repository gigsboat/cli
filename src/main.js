import path from 'path'
import marked from 'marked'

import { getAllFiles } from './utils/fs.js'
import { convertToJson } from './utils/yaml-parser.js'
import {
  getEventsMd,
  formatToMarkdown,
  getStatsBadges
} from './utils/md-formatter.js'
import { createYearBuckets, getEventsStats } from './utils/content-manager.js'

export { formatToMarkdown, generateGigs, generateDocument }

async function generateDocument({ sourceDirectory, preContent, postContent }) {
  let markdownOutputPreContent = ''
  let markdownOutputPostContent = ''
  let document = ''

  const { entriesForYearMarkdown, entriesByBucket } = await generateGigs({
    sourceDirectory
  })

  if (preContent) {
    markdownOutputPreContent = processCustomContent({
      contents: preContent
    })
  }

  if (postContent) {
    markdownOutputPostContent = processCustomContent({
      contents: postContent
    })
  }

  const statsHeader = `<div align='center'>${marked.parse(
    getStatsBadges(entriesByBucket)
  )}</div>
  `

  const currentDate = new Date().toISOString()
  const footer = `<i>Updated on ${currentDate}</i>`

  document +=
    statsHeader +
    markdownOutputPreContent +
    '\n' +
    entriesForYearMarkdown +
    '\n' +
    markdownOutputPostContent +
    '\n' +
    footer

  return document
}

function processCustomContent({ contents }) {
  let markdownContent = ''
  for (const content of contents) {
    if (content.hasOwnProperty('raw')) {
      markdownContent += content.raw + '\n'
    }

    if (content.hasOwnProperty('format')) {
      markdownContent += formatToMarkdown(content.format) + '\n'
    }
  }

  return markdownContent
}

async function generateGigs({ sourceDirectory }) {
  let directoryPath = path.join(process.cwd(), sourceDirectory)
  if (path.isAbsolute(sourceDirectory)) {
    directoryPath = sourceDirectory
  }

  const allFilesFlatList = await getAllFiles(directoryPath)

  const entries = []
  for (const filePath of allFilesFlatList) {
    const json = await convertToJson(filePath)

    // push the relative path to the file as part of the data object
    let fileDirectoryPath = path.relative(directoryPath, filePath)
    if (!path.isAbsolute(sourceDirectory)) {
      fileDirectoryPath = path.join(sourceDirectory, fileDirectoryPath)
    }
    json.fileRelativePath = fileDirectoryPath

    entries.push(json)
  }

  const entriesByBucket = await getEntriesByBuckets(entries)
  const entriesForYearMarkdown = getEventsMd(entriesByBucket.bucketsByYear)
  return {
    entriesForYearMarkdown,
    entriesByBucket
  }
}

async function getEntriesByBuckets(entries) {
  // @TODO these buckets:
  //  const bucketsLanguage = {}
  //  const bucketsCountry = {}

  const bucketsByYear = createYearBuckets(entries)
  const statsTotal = getEventsStats(entries)

  return {
    bucketsByYear,
    stats: statsTotal
  }
}
