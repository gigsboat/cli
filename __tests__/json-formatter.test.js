import { jest } from '@jest/globals'
import path from 'path'
import { readFile } from 'fs/promises'

import { generateDocument } from '../src/main.js'

const __dirname = new URL('.', import.meta.url).pathname

beforeEach(() => {
  jest.useFakeTimers('modern')
  jest.setSystemTime(new Date(Date.UTC(2022, 11, 14)))
})

afterAll(() => jest.useRealTimers())

test('given a directory of markdown files ensure the JSON output is created', async () => {
  const filePath = path.join(__dirname, './__fixtures__/main-datafiles')
  await generateDocument({ sourceDirectory: filePath })

  const jsonStringData = await readFile(path.join('README.json'), 'utf8')
  const jsonResult = JSON.parse(jsonStringData)
  expect(jsonResult).toMatchSnapshot()
})