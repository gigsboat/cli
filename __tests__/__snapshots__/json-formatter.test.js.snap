// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a directory of markdown files a full document is rendered 1`] = `
"{
  \\"items\\": [
    {
      \\"date\\": \\"2021-07-21T00:00:00.000Z\\",
      \\"tags\\": \\"post\\",
      \\"name\\": \\"Cyber Week\\",
      \\"url\\": \\"https://cw2021.b2b-wizard.com/expo\\",
      \\"type\\": \\"conference\\",
      \\"title\\": \\"Are We Forever Doomed By Software Supply Chain Risks?\\",
      \\"slides_url\\": null,
      \\"recording_url\\": \\"https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk\\",
      \\"city\\": \\"Tel Aviv\\",
      \\"country\\": \\"Israel\\",
      \\"country_code\\": \\"IL\\",
      \\"language\\": \\"English\\",
      \\"recognitions\\": {
        \\"twitter\\": [
          \\"https://twitter.com/liran_tal/status/1417874639859109894\\"
        ]
      },
      \\"image_header\\": null,
      \\"images\\": [
        null
      ]
    },
    {
      \\"date\\": \\"2021-06-13T00:00:00.000Z\\",
      \\"tags\\": \\"post\\",
      \\"name\\": \\"Cyber Week\\",
      \\"url\\": \\"https://cw2021.b2b-wizard.com/expo\\",
      \\"type\\": \\"article\\",
      \\"title\\": \\"Are We Forever Doomed By Software Supply Chain Risks?\\",
      \\"slides_url\\": null,
      \\"recording_url\\": \\"https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk\\",
      \\"city\\": \\"Tel Aviv\\",
      \\"country\\": \\"Israel\\",
      \\"country_code\\": \\"IL\\",
      \\"language\\": \\"English\\",
      \\"recognitions\\": {
        \\"twitter\\": [
          \\"https://twitter.com/liran_tal/status/1417874639859109894\\"
        ]
      },
      \\"image_header\\": null,
      \\"images\\": [
        null
      ]
    },
    {
      \\"date\\": \\"2021-01-01T00:00:00.000Z\\",
      \\"tags\\": \\"post\\",
      \\"name\\": \\"Cyber Week\\",
      \\"url\\": \\"https://cw2021.b2b-wizard.com/expo\\",
      \\"type\\": \\"other\\",
      \\"title\\": \\"Are We Forever Doomed By Software Supply Chain Risks?\\",
      \\"slides_url\\": null,
      \\"recording_url\\": \\"https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk\\",
      \\"city\\": \\"Tel Aviv\\",
      \\"country\\": \\"Israel\\",
      \\"country_code\\": \\"IL\\",
      \\"language\\": \\"English\\",
      \\"recognitions\\": {
        \\"twitter\\": [
          \\"https://twitter.com/liran_tal/status/1417874639859109894\\"
        ]
      },
      \\"image_header\\": null,
      \\"images\\": [
        null
      ]
    },
    {
      \\"date\\": \\"2019-09-27T00:00:00.000Z\\",
      \\"tags\\": \\"post\\",
      \\"name\\": \\"JSConf Budapest\\",
      \\"url\\": \\"https://2019.jsconfbp.com/\\",
      \\"type\\": \\"meetup\\",
      \\"title\\": \\"StrangerDanger: Finding Security Vulnerabilities Before They Find You!\\",
      \\"slides_url\\": null,
      \\"recording_url\\": \\"https://www.youtube.com/watch?v=3H8pF6yoSgU&list=PL37ZVnwpeshEMCvdYDdZ09Sy-toTftWu0&index=15&ab_channel=JSConf\\",
      \\"city\\": \\"Budapest\\",
      \\"country\\": \\"Hungary\\",
      \\"country_code\\": \\"HU\\",
      \\"language\\": \\"English\\",
      \\"recognitions\\": {
        \\"twitter\\": [
          \\"https://twitter.com/panay_georgiou/status/1177572765664194560\\"
        ]
      },
      \\"image_header\\": \\"https://pbs.twimg.com/media/EFeULhIW4AARwZa?format=jpg&name=large\\",
      \\"images\\": [
        \\"https://pbs.twimg.com/media/EFeULhIW4AARwZa?format=jpg&name=large\\"
      ]
    },
    {
      \\"date\\": \\"2019-09-15T00:00:00.000Z\\",
      \\"tags\\": \\"post\\",
      \\"name\\": \\"JSConf Budapest\\",
      \\"url\\": \\"https://2019.jsconfbp.com/\\",
      \\"type\\": \\"podcast\\",
      \\"title\\": \\"StrangerDanger: Finding Security Vulnerabilities Before They Find You!\\",
      \\"slides_url\\": null,
      \\"recording_url\\": \\"https://www.youtube.com/watch?v=3H8pF6yoSgU&list=PL37ZVnwpeshEMCvdYDdZ09Sy-toTftWu0&index=15&ab_channel=JSConf\\",
      \\"city\\": \\"Budapest\\",
      \\"country\\": \\"Hungary\\",
      \\"country_code\\": \\"HU\\",
      \\"language\\": \\"English\\",
      \\"recognitions\\": {
        \\"twitter\\": [
          \\"https://twitter.com/panay_georgiou/status/1177572765664194560\\"
        ]
      },
      \\"image_header\\": \\"https://pbs.twimg.com/media/EFeULhIW4AARwZa?format=jpg&name=large\\",
      \\"images\\": [
        \\"https://pbs.twimg.com/media/EFeULhIW4AARwZa?format=jpg&name=large\\"
      ]
    },
    {
      \\"date\\": \\"2019-09-12T00:00:00.000Z\\",
      \\"tags\\": \\"post\\",
      \\"name\\": \\"APIdays Barcelona\\",
      \\"url\\": \\"https://www.apidays.co/barcelona2019/\\",
      \\"type\\": \\"webinar\\",
      \\"title\\": \\"Consumer-Driven Contracts: A better approach for API Testing\\",
      \\"slides_url\\": \\"https://slides.com/lirantal/consumer-driven-contracts\\",
      \\"recording_url\\": \\"https://www.youtube.com/watch?v=zfGKX5iKSis&list=PLmEaqnTJ40Oqo9VlbcUakVmFZgx5weLrs&index=25&t=141s&ab_channel=apidays\\",
      \\"city\\": \\"Barcelona\\",
      \\"country\\": \\"Spain\\",
      \\"country_code\\": \\"ES\\",
      \\"language\\": \\"English\\",
      \\"recognitions\\": {
        \\"twitter\\": [
          \\"https://twitter.com/anyulled/status/1172079037189152768\\",
          \\"https://twitter.com/APIdaysGlobal/status/1162621839933358080\\",
          \\"https://twitter.com/liran_tal/status/1173942213086076928\\",
          \\"https://twitter.com/liran_tal/status/1199047628530556931\\",
          \\"https://twitter.com/snyksec/status/1172033103528898561\\"
        ]
      },
      \\"image_header\\": \\"https://pbs.twimg.com/media/EEQPqXlXUAEoafi?format=jpg&name=large\\",
      \\"images\\": [
        \\"https://pbs.twimg.com/media/EEQPqXlXUAEoafi?format=jpg&name=large\\",
        \\"https://pbs.twimg.com/media/ECJ2Zv6WsAEPUFF?format=jpg&name=large\\"
      ]
    }
  ],
  \\"stats\\": {
    \\"total\\": 6,
    \\"total_podcast\\": 1,
    \\"total_conference\\": 1,
    \\"total_webinar\\": 1,
    \\"total_meetup\\": 1,
    \\"total_article\\": 1,
    \\"total_workshop\\": 0,
    \\"total_other\\": 1
  },
  \\"generatedAt\\": \\"2022-12-14T00:00:00.000Z\\"
}"
`;
